compiler/src/dmd/location.d

```diff
@@ -38,19 +33,19 @@ debug info etc.
 */
 struct Loc
 {
-    private uint _linnum;
-    private uint _charnum;
-    private uint fileIndex; // index into filenames[], starting from 1 (0 means no filename)
-    version (LocOffset)
-        uint fileOffset; /// utf8 code unit index relative to start of file, starting from 0
+    // https://github.com/dlang/dmd/pull/20777#issuecomment-2614128849
+    static if (size_t.sizeof == 4) version(linux)
+    {
+        uint dummy;
+    }
```

Comment:
Well I found a fix :facepalm: 

Clearly the optimization is just too good for i386 Linux to handle.

Reply:
You probably want DigitalMars instead of Linux.

Reply:
The failure is specific to Ubuntu x86, Windows x86 works fine. It's also specific to DMD compiled with DMD latest, not DMD bootstrap. It happens when calling `StructDeclaration.setZeroInit(bool)`, which is a simple bit field setter. `this` is not `null`, and setting `this.bitFields = 0` doesn't trigger it, so what kind of memory access even triggers the segfault? I truly can't explain this.

Reply:
Also the stack frame is really shallow (13 frames), I can't imagine it hitting the guard page

Reply:
If this is bug in the old DMD bootstrap compiler, then please use both `version(linux)` (or Posix, I can't imagine this to be Linux specific) and `version(DigitalMars)`, so that the ugly workaround doesn't apply when using LDC or GDC (incl. x86 builds of those compilers themselves, as they hardly use DMD as a host compiler).

Reply:
Btw that's DMD v2.097.1 for the Linux x86 job; Linux x64 uses v2.079.0. So looks like there were other 32-bit problems already, requiring a newer version. Might be worth investigating if/when it was fixed at one point, and if so, bump it accordingly.

Reply:
> It's also specific to DMD compiled with DMD latest, not DMD bootstrap.

Oh sorry, I misread that - so even worse, a regression in recent versions?!

Reply:
https://github.com/dlang/dmd/pull/20786 indicates that the regression was introduced in DMD v2.105.

Reply:
Thanks for the investigation! 

> `version(DigitalMars)`

How do I put that in the C++ header?

Reply:
By chance, I do have a debian 12 x86 VM to hand, and couldn't reproduce the segfault with the host compiler being either 2.104, 2.105, or 2.109.

Not sure it's really worth checking an ubuntu 22.04 x86 VM set up.

Reply:
> How do I put that in the C++ header?

See dcompat.h

https://github.com/dlang/dmd/blob/master/compiler/src/dmd/root/dcompat.h#L42-L43